File: ruby/064_Uplift.rb
# 064 - Uplift（★3）
# 階差

N, Q = gets.split.map(&:to_i)
A = gets.split.map(&:to_i)
LRV = Array.new(Q) { gets.split.map(&:to_i) }
# 最初の階差を計算
diffs = (1...N).map { A[_1] - A[_1 - 1] }
# 最初の不便さを計算
inconvenience = diffs.sum(&:abs)

ans = []
LRV.each do |l, r, v|
  l, r = l - 1, r - 1
  # 影響を受けるのは境界の階差のみ
  if l > 0
    pre_diff = diffs[l - 1]
    diffs[l - 1] += v
    inconvenience += diffs[l - 1].abs - pre_diff.abs
  end
  if r < N - 1
    pre_diff = diffs[r]
    diffs[r] -= v
    inconvenience += diffs[r].abs - pre_diff.abs
  end
  ans << inconvenience
end

puts ans

File: kotlin/064_Uplift.kt
/**
 * 064 - Uplift（★3）
 * 階差
 */

 fun main() {
  val (N, Q) = readLine()!!.split(" ").map(String::toInt)
  val A = readLine()!!.split(" ").map(String::toLong)
  val LRV = Array(Q) { readLine()!!.split(" ").map(String::toInt) }
  val diffs = (1 .. N - 1).map { A[it] - A[it - 1] }.toTypedArray()
  // sumByの戻り値はInt型になるためsumByは使用せず
  var inconvinience = diffs.map { Math.abs(it) }.sum()

  val ans = LongArray(Q)

  repeat(Q) { i ->
    val l = LRV[i][0] - 1
    val r = LRV[i][1] - 1
    val v = LRV[i][2].toLong()
    if (l > 0) {
      val preDiff = diffs[l - 1]
      diffs[l - 1] += v
      inconvinience += Math.abs(diffs[l - 1]) - Math.abs(preDiff)
    }
    if (r < N - 1) {
      val preDiff = diffs[r]
      diffs[r] -= v
      inconvinience += Math.abs(diffs[r]) - Math.abs(preDiff)
    }
    ans[i] = inconvinience

  }
  ans.forEach(::println)
}
